src/main/java/org/opensearch/neuralsearch/processor/NeuralSparseTwoPhaseProcessor.java

```diff
@@ -0,0 +1,230 @@
+
+            boolean enabled = ConfigurationUtils.readBooleanProperty(TYPE, tag, config, ENABLE_KEY, true);
+            Map<String, Object> map = ConfigurationUtils.readOptionalMap(TYPE, tag, config, PARAMETER_KEY);
+
+            float ratio = 0.4f;
+            float window_expansion = 5.0f;
+            int max_window_size = 10000;
+            if (map != null) {
+                ratio = ((Number) map.getOrDefault(RATIO_KEY, ratio)).floatValue();
+                window_expansion = ((Number) map.getOrDefault(EXPANSION_KEY, window_expansion)).floatValue();
+                max_window_size = ((Number) map.getOrDefault(MAX_WINDOW_SIZE_KEY, max_window_size)).intValue();
+            }
+
+            return new NeuralSparseTwoPhaseProcessor(tag, description, ignoreFailure, enabled, ratio, window_expansion, max_window_size);
+        }
+    }
+
+    private QueryBuilder getNestedQueryBuilderFromNeuralSparseQueryBuilderMap(
+        final Multimap<NeuralSparseQueryBuilder, Float> queryBuilderFloatMap
+    ) {
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        queryBuilderFloatMap.asMap().forEach((neuralSparseQueryBuilder, boosts) -> {
+            float reduceBoost = boosts.stream().reduce(0.0f, Float::sum);
+            final float finalBoost = neuralSparseQueryBuilder.boost() * reduceBoost;
+            boolQueryBuilder.should(neuralSparseQueryBuilder.boost(finalBoost));
+        });
+        return boolQueryBuilder;
+    }
+
+    private float getOriginQueryWeightAfterRescore(final SearchSourceBuilder searchSourceBuilder) {
+        if (searchSourceBuilder.rescores() == null) return 1.0f;
+        return searchSourceBuilder.rescores()
+            .stream()
+            .map(rescorerBuilder -> ((QueryRescorerBuilder) rescorerBuilder).getQueryWeight())
+            .reduce(1.0f, (a, b) -> a * b);
+    }
+
+    private void collectNeuralSparseQueryBuilder(
+        final QueryBuilder queryBuilder,
+        final Multimap<NeuralSparseQueryBuilder, Float> neuralSparseQueryBuilderFloatMap,
+        float baseBoost
+    ) {
+        if (queryBuilder instanceof BoolQueryBuilder) {
+            collectNeuralSparseQueryBuilder((BoolQueryBuilder) queryBuilder, neuralSparseQueryBuilderFloatMap, baseBoost);
+        } else if (queryBuilder instanceof NeuralSparseQueryBuilder) {
+            collectNeuralSparseQueryBuilder((NeuralSparseQueryBuilder) queryBuilder, neuralSparseQueryBuilderFloatMap, baseBoost);
+        }
+    }
+
+    private void collectNeuralSparseQueryBuilder(
+        final BoolQueryBuilder queryBuilder,
+        final Multimap<NeuralSparseQueryBuilder, Float> neuralSparseQueryBuilderFloatMap,
+        float baseBoost
+    ) {
+        baseBoost *= queryBuilder.boost();
+        for (QueryBuilder subQuery : queryBuilder.should()) {
+            if (subQuery instanceof BoolQueryBuilder) {
+                collectNeuralSparseQueryBuilder(subQuery, neuralSparseQueryBuilderFloatMap, baseBoost);
```

Comment:
which one of your two methods will be called here? I think for clarity it's better to rename them, so it's clear from the name if it accepts BooleanQuery or generic Query

Reply:
I integrated them into a one more simple function now.